test/apps/network/netlink_route.c

```diff
@@ -0,0 +1,222 @@
+	char buffer[BUFFER_SIZE];
+
+	// Create a new netlink socket
+	sock_fd = TEST_SUCC(socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE));
+
+	memset(&sa, 0, sizeof(sa));
+	sa.nl_family = AF_NETLINK;
+
+	// Bind the socket
+	TEST_SUCC(bind(sock_fd, (struct sockaddr *)&sa, sizeof(sa)) < 0);
+
+	// Build the request
+	struct nl_req req;
+	memset(&req, 0, sizeof(req));
+	req.hdr.nlmsg_len = NLMSG_LENGTH(sizeof(struct rtgenmsg));
+	req.hdr.nlmsg_type = RTM_GETLINK;
+	req.hdr.nlmsg_flags = NLM_F_REQUEST | NLM_F_DUMP;
+	req.hdr.nlmsg_seq = 1;
+	req.gen.rtgen_family = AF_PACKET;
+
+	// Send the request
+	struct iovec iov = { &req, req.hdr.nlmsg_len };
+	struct msghdr msg = { &sa, sizeof(sa), &iov, 1, NULL, 0, 0 };
+
+	TEST_SUCC(sendmsg(sock_fd, &msg, 0));
+
+	// Get the response
+	ssize_t len;
+	bool parse_done = false;
+	while (1) {
+		if (parse_done) {
+			break;
+		}
+
+		len = TEST_SUCC(recv(sock_fd, buffer, BUFFER_SIZE, 0));
+
+		struct nlmsghdr *nlh = (struct nlmsghdr *)buffer;
+
+		// 解析每个消息
+		for (; NLMSG_OK(nlh, len); nlh = NLMSG_NEXT(nlh, len)) {
+			if (nlh->nlmsg_type == NLMSG_DONE) {
+				parse_done = true;
+			}
+
+			if (nlh->nlmsg_type == NLMSG_ERROR) {
+				perror("netlink error");
+				parse_done = true;
+			}
+
+			if (nlh->nlmsg_type == RTM_NEWLINK) {
+				struct ifinfomsg *ifi = NLMSG_DATA(nlh);
+				struct rtattr *attr = IFLA_RTA(ifi);
+				int remaining = nlh->nlmsg_len -
+						NLMSG_LENGTH(sizeof(*ifi));
+
+				// Parse each attribute
+				for (; RTA_OK(attr, remaining);
+				     attr = RTA_NEXT(attr, remaining)) {
+					if (attr->rta_type == IFLA_IFNAME) {
+						printf("Interface: %s\n",
+						       (char *)RTA_DATA(attr));
+						break;
+					}
+				}
+			}
+		}
+	}
+
+	TEST_SUCC(close(sock_fd));
+}
+END_TEST()
+
+FN_TEST(get_addresses)
+{
+	struct nl_req {
+		struct nlmsghdr hdr;
+		struct ifaddrmsg ifa;
+	};
+
+	void parse_rtattr(struct rtattr * tb[], int max, struct rtattr *rta,
+			  ssize_t len)
+	{
+		memset(tb, 0, sizeof(struct rtattr *) * (max + 1));
+		while (RTA_OK(rta, len)) {
+			if (rta->rta_type <= max) {
+				tb[rta->rta_type] = rta;
+			}
+			rta = RTA_NEXT(rta, len);
+		}
+	}
+
+	struct sockaddr_nl sa;
+	int fd;
+	char buffer[BUFFER_SIZE];
+	struct nlmsghdr *nlh;
+	struct nl_req req;
+
+	// Create a new netlink socket
+	fd = TEST_SUCC(socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE));
+
+	// Bind the socket
+	memset(&sa, 0, sizeof(sa));
+	sa.nl_family = AF_NETLINK;
+	sa.nl_groups = 0;
+
+	TEST_SUCC(bind(fd, (struct sockaddr *)&sa, sizeof(sa)));
+
+	// Build the request
+	memset(&req, 0, sizeof(req));
+	req.hdr.nlmsg_len = NLMSG_LENGTH(sizeof(struct ifaddrmsg));
+	req.hdr.nlmsg_type = RTM_GETADDR;
+	req.hdr.nlmsg_flags = NLM_F_REQUEST | NLM_F_DUMP;
+	req.hdr.nlmsg_seq = 1;
+	req.ifa.ifa_family = AF_UNSPEC; // Get address of all families
+
+	// Send the request
+	TEST_SUCC(send(fd, &req, req.hdr.nlmsg_len, 0));
+
+	// Receive the response
+	ssize_t len;
+	bool parse_done = false;
+	while (1) {
+		if (parse_done) {
```

Comment:
I suppose [`libnl`](https://www.infradead.org/~tgr/libnl/) can be used to simplify the regression tests. For example, the link/interface information is much easier to get using the [`rtnl_link_get_xxx` APIs](https://www.infradead.org/~tgr/libnl/doc/api/group__link.html).